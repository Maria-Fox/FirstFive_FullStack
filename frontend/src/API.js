import axios from "axios";

const BASE_URL = process.env.REACT_APP_BASE_URL || "https://localhost:3001";

class API {

  // Used to authenticate user
  static token = "";

  // ******************************************* API send request method

  static async request(endpoint, data = {}, method = "get") {
    // console.debug("API Call:", endpoint, data, method, this.token, "**********");

    if (!this.token) {
      this.token = window.localStorage.getItem("token") || "invalidTemp"
    };

    const url = `${BASE_URL}/${endpoint}`;
    const headers = { Authorization: `Bearer ${API.token}`, 'Access-Control-Allow-Origin': 'https://firstfive.onrender.comtps://firstfive.surge.sh'};
    const params = (method === "get")
      ? data
      : {};


    try {
      return (await axios({ url, method, data, params, headers })).data;
    } catch (err) {
      // console.error("API Error:", err.message);
      // grab just the error message
      let message = err.response.data.error.message;
      throw Array.isArray(message) ? message : [message];
    };
  };

  // ******************************************* Register/ LOGIN methods


  static async registerUser(userData) {
    let newUser = await this.request("auth/register", { ...userData }, 'post');
    this.token = newUser.signedJWT
    return newUser.signedJWT;
  };

  static async authenticateUser(userData) {
    let authUser = await this.request(`auth/login`, { ...userData }, 'post');
    return authUser.signedJWT;
  };




  // ******************************************* USER methods

  // Need to see about limmited views to only those who matched..


  static async viewUser(username) {
    let res = await this.request(`users/${username}`);
    return res.userData;
  };

  static async viewAuthUserProfile(username) {
    let res = await this.request(`users/${username}/profile`);
    return res.userData;
  };

  static async editUser(username, updateData) {
    let res = await this.request(`users/${username}`, { ...updateData }, 'patch');
    return res.userData;
  };

  static async deleteUser(username) {
    let res = await this.request(`users/${username}`, {}, 'delete');
    return res;
  };


  // ******************************************* PROJECT methods

  static async createProject(projectData) {
    let res = await this.request(`projects/add`, { ...projectData }, 'post');
    return res;
  };

  static async getAllProjects() {
    let res = await this.request(`projects/all`);
    return res;
  };

  static async getNonMatchedProjects() {
    let res = await this.request(`projects/unmatched`);
    return res;
  };

  static async carouselProjects() {
    let res = await this.request(`projects/carousel`);
    return res;
  };

  static async getUserCreatedProjects(username) {
    let res = await this.request(`projects/created/by/${username}`);
    return res;
  };

  static async viewProject(project_id) {
    let res = await this.request(`projects/${project_id}`);
    return res;
  };

  static async editProject(project_id, updateData) {
    let res = await this.request(`projects/${project_id}`, { ...updateData }, 'patch');
    return res;
  };

  static async deleteProject(project_id) {
    let res = await this.request(`projects/${project_id}`, {}, 'delete');
    return res;
  };

  // ******************************************* Match methods

  static async addMatch(username, project_id) {
    let res = await this.request(`matches/${username}/${project_id}`, {}, 'post');
    return res;
  };

  static async viewUserMatches(username) {
    let res = await this.request(`matches/${username}/all`);
    return res;
  };

  static async viewProjectUserMatches(project_id) {
    let res = await this.request(`matches/${project_id}/users`);
    return res;
  };

  static async removeUserMatch(username, project_id) {
    let res = await this.request(`matches/${username}/${project_id}`, {}, 'delete');
    return res;
  };


  // ******************************************* Project Member methods

  static async addProjectMember(project_id, username) {
    let res = await this.request(`projectmembers/add/${project_id}`, { username }, 'post');
    return res;
  };

  static async viewAllProjMembers(project_id) {
    let res = await this.request(`projectmembers/${project_id}/users`);
    return res;
  };

  static async deleteProjectMember(project_id, username) {
    let res = await this.request(`projectmembers/${project_id}`, { username }, 'delete');
    return res;
  };

  // ******************************************* Message methods

  static async createMessage(username, msgData) {
    let res = await this.request(`messages/${username}/create`, { ...msgData }, 'post');
    return res;
  };

  static async getAllUserMessages(username) {
    let res = await this.request(`messages/${username}/all`);
    return res;
  };

  static async viewSingleMsgData(message_id, username) {
    let res = await this.request(`messages/${username}/read/${message_id}`);
    return res;
  };


  // class closing bracket don't delete.
};

export default API;