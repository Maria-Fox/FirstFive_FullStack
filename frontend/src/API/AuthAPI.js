import axios from "axios";

const BASE_URL = process.env.REACT_APP_BASE_URL || "http://localhost:3001";

class AuthAPI {

  // Used to authenticate user
  static token = "";

  // ******************************************* API send request method

  static async request(endpoint, data = {}, method = "get") {

    if (!this.token) {
      this.token = window.localStorage.getItem("token") || "invalidTemp"
    };

    const url = `${BASE_URL}/${endpoint}`;
    const headers = { Authorization: `Bearer ${API.token}` };
    const params = (method === "get")
      ? data
      : {};


    try {
      return (await axios({ url, method, data, params, headers })).data;
    } catch (err) {
      console.error("API Error:", err.message);
      // grab just the error message
      let message = err.response.data.error.message;
      throw Array.isArray(message) ? message : [message];
    };
  };


  // ******************************************* Register/ LOGIN methods


  static async registerUser(userData) {
    let newUser = await this.request("auth/register", { ...userData }, 'post');
    this.token = newUser.signedJWT
    return newUser.signedJWT;
  };

  static async authenticateUser(userData) {
    let authUser = await this.request(`auth/login`, { ...userData }, 'post');
    return authUser.signedJWT;
  };

};

export default AuthAPI;