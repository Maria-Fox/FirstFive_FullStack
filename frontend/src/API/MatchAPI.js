import axios from "axios";

const BASE_URL = process.env.REACT_APP_BASE_URL || "http://localhost:3001";

class MatchAPI {

  // Used to authenticate user. Sent in request header.
  static token = "";

  // ******************************************* API send request method

  static async request(endpoint, data = {}, method = "get") {

    if (!this.token) {
      this.token = window.localStorage.getItem("token") || "invalidTemp"
    };

    const url = `${BASE_URL}/${endpoint}`;
    const headers = { Authorization: `Bearer ${API.token}` };
    const params = (method === "get")
      ? data
      : {};


    try {
      return (await axios({ url, method, data, params, headers })).data;
    } catch (err) {
      console.error("API Error:", err.message);
      // grab just the error message
      let message = err.response.data.error.message;
      throw Array.isArray(message) ? message : [message];
    };
  };


  // ******************************************* Match methods

  static async addMatch(username, project_id) {
    let res = await this.request(`matches/${username}/${project_id}`, {}, 'post');
    return res;
  };

  static async viewUserMatches(username) {
    let res = await this.request(`matches/${username}/all`);
    return res;
  };

  static async removeUserMatch(username, project_id) {
    let res = await this.request(`matches/${username}/${project_id}`, {}, 'delete');
    return res;
  };

  static async viewProjectUserMatches(project_id) {
    let res = await this.request(`matches/${project_id}/users`);
    return res;
  };


};

export default MatchAPI;