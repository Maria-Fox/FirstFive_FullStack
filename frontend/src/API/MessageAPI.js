import axios from "axios";

const BASE_URL = process.env.REACT_APP_BASE_URL || "http://localhost:3001";

class MessageAPI {

  // Used to authenticate user
  static token = "";

  // ******************************************* API send request method

  static async request(endpoint, data = {}, method = "get") {

    if (!this.token) {
      this.token = window.localStorage.getItem("token") || "invalidTemp"
    };

    const url = `${BASE_URL}/${endpoint}`;
    const headers = { Authorization: `Bearer ${API.token}` };
    const params = (method === "get")
      ? data
      : {};


    try {
      return (await axios({ url, method, data, params, headers })).data;
    } catch (err) {
      console.error("API Error:", err.message);
      // grab just the error message
      let message = err.response.data.error.message;
      throw Array.isArray(message) ? message : [message];
    };
  };


  // ******************************************* Message methods

  static async createMessage(username, msgData) {
    let res = await this.request(`messages/${username}/create`, { ...msgData }, 'post');
    return res;
  };

  static async getAllUserMessages(username) {
    let res = await this.request(`messages/${username}/all`);
    return res;
  };

  static async viewSingleMsgData(message_id, username) {
    let res = await this.request(`messages/${username}/read/${message_id}`);
    return res;
  };

};

export default MessageAPI;