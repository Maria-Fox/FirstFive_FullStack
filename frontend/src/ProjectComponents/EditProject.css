#approved{
  max-width: 20vw;
  max-height: 15vh;
  color: whitesmoke;
  background-color: darkgreen;
  display: inline-block;
};

#notApproved{
  max-width: 20vw;
  max-height: 15vh;
  color: whitesmoke;
  background-color: darkred;
  display: inline-block;
}