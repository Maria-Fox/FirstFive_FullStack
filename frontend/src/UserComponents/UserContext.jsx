import React from "react";

// Create user context instance to later provide in App.js and use throughout application. Holds auth user

const UserContext = React.createContext();

export default UserContext;