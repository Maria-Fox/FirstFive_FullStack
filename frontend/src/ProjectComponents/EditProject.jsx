import React, { useState, useContext, useEffect } from "react";
import UserContext from "../UserComponents/UserContext";
import { useNavigate } from "react-router-dom";
import { useParams } from "react-router-dom";
import API from "../API";
import AlertNotification from "../Common/AlertNotifications";
import { FormGroup, Form, Card, Label, Input, Button } from "reactstrap";
import "./EditProject.css";
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { regular } from '@fortawesome/fontawesome-svg-core/import.macro'; // <-- import styles to be used


const EditProject = () => {

  // ***************************************************************


  const { id } = useParams();
  const { authUser } = useContext(UserContext);
  const navigate = useNavigate();

  // ***************************************************************


  let project_values = {
    "owner_username": "",
    "name": "",
    "project_desc": "",
    "timeframe": "",
    "github_repo": ""
  }

  const [projData, setProjData] = useState(project_values);
  const [errors, setErrors] = useState(null);
  const [acknowledgeUserChange, setAcknowledgeUserChange] = useState(false);


  // ***************************************************************


  useEffect(() => {
    async function preloadProjData() {
      try {
        let response = await API.viewProject(id);
        setProjData(
          {
            "name": response.name,
            "owner_username": response.owner_username,
            "project_desc": response.project_desc,
            "timeframe": response.timeframe,
            "github_repo": response.github_repo
          });

      } catch (e) {
        setErrors(e);
        return;
      }
    }

    preloadProjData();
  }, [setProjData, authUser, id]);

  // ***************************************************************


  let handleChange = (e) => {
    let { name, value } = e.target;
    setProjData(projData => ({
      ...projData,
      [name]: value
    }));
  };

  // ***************************************************************

  async function handleSubmit(e) {
    try {
      e.preventDefault();
      
      if(projData.owner_username !== authUser){
        if(!acknowledgeUserChange){
          console.log("says user change")
          setErrors(["There was a change to the project owner. To proceed, please acknowledge the change by clicking the box below and re-submit. If you did not mean to make changes please update the field to your current username."])
          return;
        };
      }

      await API.editProject(id, projData);
      navigate(`/projects/created/by/${authUser}`);
    } catch (e) {
      console.log("er have an err", e)
      setErrors(e);
      return;
    };
  };

    // ***************************************************************

    const confirmProjectOwnerChange = (
      <div >
      <Button id = {acknowledgeUserChange ? "approved" : "notApproved"}
        onClick={() => setAcknowledgeUserChange(currentStatus => !currentStatus)} >
          <FontAwesomeIcon icon={regular('circle-check')} />
        </Button>

      <p style={{display: "inline-block"}}>I acknowledge once I transfer ownership I will not have access to update any project details including project members.</p>
    
      </div>
    )


  // ***************************************************************

  return (
    <div>
      {!projData ? <p>Loading...</p> :
        <div className="container">
          <h1 className="text-center text-white pt-2 mt-2">Update: {projData.name}</h1>

          <Card >
            {errors ? <AlertNotification messages={errors} /> : null}

            <Form onSubmit={handleSubmit} className="p-4">

              <FormGroup>
                <Label htmlFor="name" >Project Name
                  <Input
                    type="text"
                    id="name"
                    value={projData.name}
                    name="name"
                    required
                    onChange={handleChange}
                  >
                  </Input>
                </Label>
              </FormGroup>


              <FormGroup>
                <Label htmlFor="owner_username" >Project Owner
                  <Input
                    type="text"
                    id="owner_username"
                    value={projData.owner_username}
                    name="owner_username"
                    required
                    onChange={handleChange}
                  >
                  </Input>
                </Label>
              </FormGroup>

              {projData.owner_username !== authUser ? confirmProjectOwnerChange : null}

              <FormGroup>
                <Label htmlFor="project_desc" >Project Description
                  <Input
                    type="textarea"
                    id="project_desc"
                    value={projData.project_desc}
                    name="project_desc"
                    required
                    cols="35"
                    rows="8"
                    onChange={handleChange}
                  >
                  </Input>
                </Label>
              </FormGroup>

              <FormGroup>
                <Label htmlFor="timeframe" >Timeframe
                  <Input
                    type="text"
                    id="timeframe"
                    value={projData.timeframe}
                    name="timeframe"
                    placeholder="3 weeks"
                    required
                    onChange={handleChange}
                  >
                  </Input>
                </Label>
              </FormGroup>

              <FormGroup>
                <Label htmlFor="github_repo" >Github Repo
                  <Input
                    type="text"
                    id="github_repo"
                    value={projData.github_repo}
                    name="github_repo"
                    onChange={handleChange}
                  >
                  </Input>
                </Label>
              </FormGroup>

              <Button type="submit">Submit</Button>

            </Form>
          </Card>
        </div>
      }
    </div>
  )
};

export default EditProject;