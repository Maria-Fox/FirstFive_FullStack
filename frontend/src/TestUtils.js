// Used to create sample test user data & test UserContect provider..

import React from "react";
import UserContext from "./UserComponents/UserContext";

const authUser = {
  username: "softwareDev1"
};

// sample / data
const matchedProjectIds = [1, 2, 3, 4];

const userWithMatchState = { username: "softwareDev1", matchedProjectIds: [] };

// Creating a context provider and passing in the test user, matching ids. Passing in chidlren.
const UserProvider =
  ({ children, userWithMatchState }) => {
    <UserContext.Provider value={userWithMatchState}>
      {children}
    </UserContext.Provider>
  };

export { UserProvider, authUser, userWithMatchState };