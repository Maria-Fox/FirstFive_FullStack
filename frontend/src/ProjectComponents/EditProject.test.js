import { render, waitFor } from "@testing-library/react";
import EditProject from "./EditProject";
import UserContext from "../UserComponents/UserContext";
import { MemoryRouter } from "react-router-dom";
import API from "../API";
import { authUser } from "../TestUtils";
jest.mock("../API")

describe("Rendering edit project form", function () {

  test("Smoke- Valid User- rendered edit project component.",
    function () {

      const authUser = { username: "softwareDev1" };

      render(
        <MemoryRouter>
          <UserContext.Provider value={authUser}>
            <EditProject />
          </UserContext.Provider>
        </MemoryRouter>
      );
    });

  test("Snapshot- matches project edit.", function () {

    const { asFragment } = render(
      <MemoryRouter>
        <UserContext.Provider value={{ username: "softwareDev1" }}>
          <EditProject />
        </UserContext.Provider>
      </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test("Popualtes form w/ existing projectData", function () {
    API.mockImplementationOnce(() => {
      return {
        request: () => {
          return {
            "owner_username": authUser,
            "name": "Tes name",
            "project_desc": "test desc",
            "timeframe": "1 day",
            "github_repo": "www.github.com"
          }
        }
      }
    });

    const renderedContent = render(
      <MemoryRouter>
        <UserContext.Provider value={authUser}>
          <EditProject />
        </UserContext.Provider>
      </MemoryRouter>
    );

    waitFor(() => {
      expect(renderedContent.getByText("1 day")).toBeInTheDocument();
    })

  })
});

