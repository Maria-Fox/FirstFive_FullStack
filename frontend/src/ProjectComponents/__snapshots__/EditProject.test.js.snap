// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering edit project form Snapshot- matches project edit. 1`] = `
<DocumentFragment>
  <div>
    <div
      class="container"
    >
      <h1
        class="text-center text-white pt-2 mt-2"
      >
        Update: 
      </h1>
      <div
        class="card"
      >
        <form
          class="p-4"
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="name"
            >
              Project Name
              <input
                class="form-control"
                id="name"
                name="name"
                required=""
                type="text"
                value=""
              />
            </label>
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="owner_username"
            >
              Project Owner
              <input
                class="form-control"
                id="owner_username"
                name="owner_username"
                required=""
                type="text"
                value=""
              />
            </label>
          </div>
          <div>
            <button
              class="btn btn-secondary"
              id="notApproved"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-circle-check "
                data-icon="circle-check"
                data-prefix="far"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M243.8 339.8C232.9 350.7 215.1 350.7 204.2 339.8L140.2 275.8C129.3 264.9 129.3 247.1 140.2 236.2C151.1 225.3 168.9 225.3 179.8 236.2L224 280.4L332.2 172.2C343.1 161.3 360.9 161.3 371.8 172.2C382.7 183.1 382.7 200.9 371.8 211.8L243.8 339.8zM512 256C512 397.4 397.4 512 256 512C114.6 512 0 397.4 0 256C0 114.6 114.6 0 256 0C397.4 0 512 114.6 512 256zM256 48C141.1 48 48 141.1 48 256C48 370.9 141.1 464 256 464C370.9 464 464 370.9 464 256C464 141.1 370.9 48 256 48z"
                  fill="currentColor"
                />
              </svg>
            </button>
            <p
              style="display: inline-block;"
            >
              I acknowledge once I transfer ownership I will not have access to update any project details including project members.
            </p>
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="project_desc"
            >
              Project Description
              <textarea
                class="form-control"
                cols="35"
                id="project_desc"
                name="project_desc"
                required=""
                rows="8"
              />
            </label>
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="timeframe"
            >
              Timeframe
              <input
                class="form-control"
                id="timeframe"
                name="timeframe"
                placeholder="3 weeks"
                required=""
                type="text"
                value=""
              />
            </label>
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="github_repo"
            >
              Github Repo
              <input
                class="form-control"
                id="github_repo"
                name="github_repo"
                type="text"
                value=""
              />
            </label>
          </div>
          <button
            class="btn btn-secondary"
            type="submit"
          >
            Submit
          </button>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
